Split conversion and defaulting

smarterclayton opened the pull request:
Separate conversion and defaulting.   Defaulting occurs mixed with conversion today - change the server so that the `VersioningCodec` performs defaulting on the external type during decoding.  
- Add a new method to `Scheme` - `func (*runtime.Scheme) Default(runtime.Object)` - that takes an object and performs defaulting.  
- Call `Default` during decoding and at static initialization time
- Use the new `defaulter-gen` to generate top level object defaulters (`v1.Pod`) at build time for any type that needs to perform defaulting.  
- Add tests and alter the existing code to adapt as necessary
- Fix a few bugs in conversions that depended on defaulting behavior

---

Step 1 of decoupling conversion and defaulting. The generator will assist in creating top level defaulters that in a single method invoke all nested defaulters, preventing the need to recurse via reflection or conversion. These top level defaulters will be registered in the scheme and invoked instead of the nested recursion path. This will set the stage for a future generator, capable of creating defaulters from embedded struct tags on external types. However, we must gradually switch these over.

The immediate goal here is to split defaulting and conversion so that the unsafe convertor can be used to maximum potential (we would be able to use direct memory conversion for any identical nested struct, even those that must be defaulted).

The generator uses `k8s:defaulter-gen=TypeMeta` on most public packages to flag any top level type that has defaulters to get a `SetObjectDefaults_NAME` function created (types that don't have defaulters won't have functions).  This also creates a `RegisterDefaults` method that applies a default to an interface{} and returns true if the object was handled.  Existing defaults are left as is.

Add a test to verify old and new path generate the same outcomes.  Defaulter will move to gengo before this is merged, and subsequent PRs will remove defaulting during conversion and have the VersioningCodec apply defaults.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33352)

<!-- Reviewable:end -->

smarterclayton committed:
Create a generator for defaulters

Given an object that wishes to have a top level defaulter, traverse the
object looking for nested fields that have defaulters and generate a
single function for that type that invokes all defaulters. The function
will have the name `SetObjectDefaults_NAME`.

Types use `// +k8s:defaulter-gen=true` to indicate they wish a defaulter
generated. If a function already exists with the desired name
`SetObjectDefaults_NAME` then no generation will occur. At a package
level, authors can bulk select the types to generate by setting the
value of the comment to the name of a field - all objects with that
field name without `// +k8s:defaulter-gen=false` defined on the type
will get a defaulter.

Because the defaulting behavior from conversions happens recursively,
all defaulters are expected to be invoked. We call these defaulters
"non-covering" (other defaulters may be invoked beneath them). The
defaulters we generate, by comparison, are "covering" - no nested
defaulters should be invoked. To distinguish between these two types, we
introduce the `// +k8s:defaulter-gen=covers` comment on a defaulter
function which will instruct the generator that the function should
terminate recursion.

This sets the stage for future defaulter generation from comments by
subsuming our existing generators

smarterclayton committed:
Add a new Default() method on Scheme

Allow defaulter functions to be registered with the scheme.

smarterclayton committed:
Update makefile to generate defaulters

smarterclayton committed:
Flag all packages as needing defaulters

smarterclayton committed:
generated: Defaulters

smarterclayton committed:
Register new defaulting functions

smarterclayton committed:
Add a defaulting test that verifies old and new defaults are equivalent

smarterclayton committed:
Move old defaulters to public functions

Also make Deployment defaulting resilient to fuzzing

smarterclayton commented:
Extracted defaulter-gen into https://github.com/kubernetes/gengo/pull/10, have not rebased.

smarterclayton committed:
bump(k8s.io/gengo):4a9ebbace691333e73f9978d798b1bad6c53a50d

smarterclayton committed:
Make defaulting optional during conversion generation

Set the Kubernetes default to "no defaulting"

smarterclayton committed:
Hand generated conversion should check nil

smarterclayton committed:
generated: Remove defaulting from conversion

smarterclayton committed:
Make defaulting part of versioning codec

Most normal codec use should perform defaulting. DirectCodecs should not
perform defaulting. Update the defaulting_test to fuzz the list of known
defaulters. Use the new versioning.NewDefaultingCodec() method.

mbohlool commented:
Just adding a data point here #32824. I was thinking to have defaulters following the same model. I've not reviewed this PR. Will do and I may be able to get more details on that or somehow unite these two (or not if it does not make sense).

smarterclayton commented:
The next big performance win in conversion is splitting defaulting and
conversion.  I want to split that as cleanly as possible now (so I can go
fix the real performance problem in conversion without having defaulting
screwing me up), while leaving the door open to future generation of
defaulting.  I think generation of defaulting as you describe is valuable,
but it's too large of a change, and we'll probably still want to generate
defaulters for performance reasons on the server (which this sets a
framework for even if it does not implement it).

On Mon, Sep 26, 2016 at 5:02 PM, mbohlool <notifications@github.com> wrote:

> Just adding a data point here #32824
> <https://github.com/kubernetes/kubernetes/pull/32824>. I was thinking to
> have defaulters following the same model. I've not reviewed this PR. Will
> do and I may be able to get more details on that or somehow unite these two
> (or not if it does not make sense).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33352#issuecomment-249696543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2xxnNuTglQwCq3KGjYFWxqp4j85ks5quDLjgaJpZM4KEl32>
> .
>

smarterclayton committed:
Invoke init container defaulting in conversion

Because annotations...

Remove old references to DefaultingInterface in manual conversions.

smarterclayton commented:
Commits up to a27e1b0d16f4356b59d9b551ca1eaf6d30808b50 leave both code paths in place - the defaulting_test.go verifies defaulters are equivalent in the new and old paths via fuzzing.  The subsequent commits remove defaults from conversion and alter the defaulting_test.go to verify all expected types to trigger a defaulting via fuzzing.

smarterclayton committed:
Run defaulting on the scheduler startup

smarterclayton commented:
@k8s-bot unit test this

smarterclayton commented:
This is ready for review

Adding @wojtek-t

wojtek-t commented:
Nice (though I just took a glance).

@lavalamp - I'm happy to make the review of it - if you want me to do it, please reassign it to me.

smarterclayton commented:
integration flake is #33838

lavalamp commented:
I want to look at this but I also don't want to block it, so go for it.

wojtek-t commented on pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -102,11 +117,30 @@ func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, in
+
+		// perform defaulting if requested
+		if c.defaulter != nil {
+			// create a copy to ensure defaulting is not applied to the original versioned objects
```
I think I don't understand why we can't modify the original object. This has just been decoded here, so why we can't do it?

smarterclayton replied:
The contract of "VersionedObjects" is that we return the decoded, unmodified external version, and the internal, modified version.  In the old code (conversion+defaulting) only the second object had defaults applied.  This is preserving that behavior (various tests caught it).

wojtek-t commented on pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -102,11 +117,30 @@ func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, in
+			if isVersioned {
+				copied, err := c.copier.Copy(obj)
+				if err != nil {
+					copied = obj
```
This error handling is strange. Maybe at least some glog.Errorf() here?

wojtek-t replied:
Now I see it's pretty much copy from the below code, so maybe Errorf in both places...

wojtek-t reviewed (commented):
@smarterclayton - I only have few minor comments. Can you please take a look into tests (it seems they are unrelated, but please ensure they are :))

smarterclayton commented on pkg/runtime/scheme.go:
```diff
@@ -421,6 +426,23 @@ func (s *Scheme) AddDefaultingFuncs(defaultingFuncs ...interface{}) error {
 
+// AddTypeDefaultingFuncs registers a function that is passed a pointer to an
+// object and can default fields on the object. These functions will be invoked
+// when Default() is called. The newly added functions are invoked before older
```
Awkward wording, if you invoke this twice for the same object the later call takes effect.

smarterclayton commented on cmd/libs/go2idl/defaulter-gen/main.go:
```diff
@@ -0,0 +1,84 @@
+	customArgs := &generators.CustomArgs{
+		ExtraPeerDirs: []string{
+			"k8s.io/kubernetes/pkg/api",
+			"k8s.io/kubernetes/pkg/api/v1",
```
These are the defaults, should not be needed any more with the makefile

smarterclayton replied:
Removed

smarterclayton commented:
Comments addressed

thockin commented on Makefile.generated_files:
```diff
@@ -287,6 +287,117 @@ $(DEEPCOPY_GEN):
+gen_defaulter: $(DEFAULTER_FILES)
+	if [[ -f $(META_DIR)/$(DEFAULTER_GEN).todo ]]; then                  \
+	    ./hack/run-in-gopath.sh $(DEFAULTER_GEN)                         \
+	        --v $(KUBE_VERBOSE)                                         \
```
add `--logtostderr` - it was added to existing rules.

thockin commented on cmd/libs/go2idl/conversion-gen/generators/conversion.go:
```diff
@@ -36,6 +36,8 @@ import (
 type CustomArgs struct {
 	ExtraPeerDirs []string // Always consider these as last-ditch possibilities for conversions.
+	// SkipDefaulters indicates whether defaulter functions should be a part of conversion
+	SkipDefaulters bool
```
What is the EOL plan for this?

smarterclayton replied:
Probably 1.6, will comment.

thockin commented on Makefile.generated_files:
```diff
@@ -287,6 +287,117 @@ $(DEEPCOPY_GEN):
+#                  for having a defaulter generated
+
+# The result file, in each pkg, of defaulter generation.
+DEFAULTER_BASENAME := $(GENERATED_FILE_PREFIX)defaulter
```
minor: I'd argue for "defaults" for this file name

thockin commented on pkg/api/v1/zz_generated.defaulter.go:
```diff
@@ -0,0 +1,539 @@
+	return nil
+}
+
+func SetObjectDefaults_ConfigMap(in *ConfigMap) {
```
why are these public functions?

thockin commented on pkg/api/v1/zz_generated.defaulter.go:
```diff
@@ -0,0 +1,539 @@
+	SetDefaults_PodSpec(&in.Spec)
+	for i := range in.Spec.Volumes {
+		a := &in.Spec.Volumes[i]
+		SetDefaults_Volume(a)
```
I didn't review the whole thing, but I am unclear how this function came to be - can you summarize?  Is it looking for functions named "SetDefaults_Foo" for a field "Foo" or is is looking for a type-matching function or is it looking at a manually-registered list?

smarterclayton replied:
Should be described in the doc for the makefule.

It finds top level objects in packages that contain one or more fields by name (or have a direct annotation asking for this).  It then walks all children of that type, looking for any that have a package local function SetDefaults_TYPE(t *TYPE).  It then rolls those calls up into the new SetObjectDefaults.  

We would remove registration of SetDefaults_* on scheme in the future, because unlike conversion and deep copy there is no "default" behavior.

smarterclayton commented:
Any other structural changes / concerns / api problems / etc?

smarterclayton commented:
Updated with all comments and split the review commit changes into logical small changes again.

thockin commented on Makefile.generated_files:
```diff
@@ -287,6 +287,118 @@ $(DEEPCOPY_GEN):
+#       true:  always generate a defaulter for this type
+#       false: never generate a defaulter for this type
+#     on functions:
+#       covers: if the function name matches SetDefault_NAME, instructs
```
What does "covers" mean?  It feels like a random word..

smarterclayton replied:
Not random - used elsewhere in the code to describe the distinction between:

1. A defaulting function that only defaults a particular struct but not its children (non-covering)
2. A defaulting function that "covers" defaulting for itself and all nested children.

Our existing defaulting functions are non-covering - you have to call them and all nested functions to be safe.  The new generated ObjectDefaulters are covering - you don't have to call nested functions.  Some of our top level ObjectDefaulters call functions that are covering as well (other ObjectDefaulters).  I wanted to enable the specific use case of someone writing a custom defaulter that completely handles the types underneath it (overriding the generic defaulters defined in other packages).

thockin commented on cmd/libs/go2idl/defaulter-gen/main.go:
```diff
@@ -0,0 +1,78 @@
+//
+// Given a list of input directories, it will scan for top level types
+// and generate efficient defaulters for an entire object from the sum
+// of the SetDefault_* methods contained in the object tree.
```
I really dislike the convention of relying on the name to match.  I know we don't really do methods on API types, but this seems like EXACTLY what methods are for...  Rather than having to provide a public function with an obscure and non-idiomatic name, with an exact signature, a method seems vastly cleaner

wojtek-t replied:
I agree in principle, but we are already doing the same for conversions and defaulting. And changing this would probably be quite a huge change, so maybe we should fix it later?

smarterclayton replied:
This is just continuing existing code.  I know we like to rail at our past selves for bad decisions, but I'm just avoiding refactoring the other half of the code that isn't (by far) the worst problem we have in generation.

wojtek-t commented on cmd/libs/go2idl/defaulter-gen/main.go:
```diff
@@ -0,0 +1,78 @@
+limitations under the License.
+*/
+
+// defaulter-gen is a tool for auto-generating Defaulter functions.
```
s/Defaulter/defaulting/ ? (it seems that in most places we call them "defaulting functions"

wojtek-t reviewed (commented):
I don't have any further concerns. Though we should wait for @thockin to also say "ok", I guess.

thockin commented:
LGTM

smarterclayton commented:
20% more performance in conversion, here we come

wojtek-t commented:
@smarterclayton - the PR failed with panic in: k8s.io/kubernetes/pkg/api TestDefaulting 0.00s" - this seems related to your PR.

smarterclayton commented:
New type defaulter added in master, regenerated, reapplying label.  Test found a problem correctly (yay)

smarterclayton commented:
@k8s-bot unit test this

smarterclayton commented:
@k8s-bot test this issue #IGNORE

smarterclayton commented:
Rebased with fixed generations for changes, reapplying

wojtek-t commented:
@smarterclayton - is it ready for merge? Did you already understand the failure?

smarterclayton commented:
The failure looks like a flake in the makefile, where the condition of the directory the integration job is run in (after make generated_files) causes the files to be run in an incorrect order.  I'm trying to determine that that's the case today and then change the makefile for the fix.  Also: complex makefiles are not worth the time investment to save 2 seconds.

wojtek-t commented:
@smarterclayton 
```
W1018 09:12:06.533] pkg/api/defaulting_test.go:46: undefined: v2alpha1.SetObjectDefaults_JobTemplate
```
and
```
I1018 09:30:32.904] [0;31mFAILED[0m   hack/make-rules/../../hack/verify-godeps.sh	482s
```

Can you please fix?

smarterclayton commented:
That's the problem that I am debugging - the makefile is creating incorrect generations that aren't stable, and I'm trying to figure out why the generation isn't stable.

smarterclayton commented:
@k8s-bot gci gke smoke test this

smarterclayton commented:
Ok, fixed it.  I was using the wrong separator when I cargo-cult'ed the Makefile.  I replaced an xargs with an identical sed transform from later in the same file.

smarterclayton commented:
Reapplying label.

wojtek-t commented:
Cool - thanks!

mtaufen commented:
Just saw this, thanks Clayton this is great stuff!
